pinot-plugins/pinot-minion-tasks/pinot-minion-builtin-tasks/src/main/java/org/apache/pinot/plugin/minion/tasks/segmentrefresh/SegmentRefreshTaskExecutor.java

```diff
@@ -0,0 +1,200 @@
+import org.apache.pinot.segment.spi.store.SegmentDirectory;
+import org.apache.pinot.spi.config.table.TableConfig;
+import org.apache.pinot.spi.data.FieldSpec;
+import org.apache.pinot.spi.data.Schema;
+import org.apache.pinot.spi.env.PinotConfiguration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+
+public class SegmentRefreshTaskExecutor extends BaseSingleSegmentConversionExecutor {
+  private static final Logger LOGGER = LoggerFactory.getLogger(SegmentRefreshTaskGenerator.class);
+
+  private long _taskStartTime;
+
+  /**
+   * The code here currently covers segmentRefresh for the following cases:
+   * 1. Process newly added columns.
+   * 2. Addition/removal of indexes.
+   * 3. Compatible datatype change for existing columns
+   */
+  @Override
+  protected SegmentConversionResult convert(PinotTaskConfig pinotTaskConfig, File indexDir, File workingDir)
+      throws Exception {
+    _eventObserver.notifyProgress(pinotTaskConfig, "Refreshing segment: " + indexDir);
+
+    // We set _taskStartTime before fetching the tableConfig. Task Generation relies on tableConfig/Schema updates
+    // happening after the last processed time. So we explicity use the timestamp before fetching tableConfig as the
+    // processedTime.
+    _taskStartTime = System.currentTimeMillis();
+    Map<String, String> configs = pinotTaskConfig.getConfigs();
+    String tableNameWithType = configs.get(MinionConstants.TABLE_NAME_KEY);
+    String segmentName = configs.get(MinionConstants.SEGMENT_NAME_KEY);
+    String taskType = pinotTaskConfig.getTaskType();
+
+    LOGGER.info("Starting task: {} with configs: {}", taskType, configs);
+
+    TableConfig tableConfig = getTableConfig(tableNameWithType);
+    Schema schema = getSchema(tableNameWithType);
+
+    IndexLoadingConfig indexLoadingConfig = new IndexLoadingConfig(tableConfig, schema);
+    SegmentMetadataImpl segmentMetadata = new SegmentMetadataImpl(indexDir);
+    PinotConfiguration segmentDirectoryConfigs = indexLoadingConfig.getSegmentDirectoryConfigs();
+    SegmentDirectoryLoaderContext segmentLoaderContext =
+        new SegmentDirectoryLoaderContext.Builder().setTableConfig(indexLoadingConfig.getTableConfig())
+            .setSchema(schema)
+            .setInstanceId(indexLoadingConfig.getInstanceId())
+            .setSegmentName(segmentMetadata.getName())
+            .setSegmentCrc(segmentMetadata.getCrc())
+            .setSegmentDirectoryConfigs(segmentDirectoryConfigs)
+            .build();
+    SegmentDirectory segmentDirectory =
+        SegmentDirectoryLoaderRegistry.getDefaultSegmentDirectoryLoader().load(indexDir.toURI(), segmentLoaderContext);
+
+    // BaseDefaultColumnHandler part of needPreprocess() does not process any changes to existing columns like datatype,
+    // change from dimension to metric, etc.
+    boolean needPreprocess = ImmutableSegmentLoader.needPreprocess(segmentDirectory, indexLoadingConfig, schema);
```

Comment:
Do we plan to handle transformations/filters (say new derived column gets added and has a new transform function), repartitioning etc, in which case we will need something like SegmentProcessorFramework

Reply:
That's a great point. We definitely want to support additional operations like repartitioning in the future. 

However, at Linkedin, are pushes (backfills, periodic pushes) are **not** segment name agnostic. We want to make sure that the output segments exactly match the input segments in terms of name and their records. That's the reasoning for using a single segment per task and not using the segmentProcessorFramework. 

Once we move away from relying on segment names, we can followup with a code change to process multiple segments per task and also use the SegmentProcessorFramework.